test: Add nightly policy stress test

jrajahalme opened the pull request:
Add a nightly policy stress test using the connectivity-check as the
base deployment, but using CNPs that exercise the policy computation
while thousands of security identitites are added to the cluster via
`cilium kvstore` commands.

Currently kubectl port-forward must be manually run to forward the Grafana port from the test cluster:

      kubectl -n cilium-monitoring port-forward service/grafana --address 0.0.0.0 --address :: 3000:3000

Related: #11197

jrajahalme committed:
test: Do not use 2379 for stateless-etcd

etcd-k8s1 pod already uses the standard etcd port (2379), so deploying
the stateless-etcd manifest succeeded only when it happened to be
scheduled on another node. Change the port to avoid port collision
with etcd-k8s1.

This fixes likely test flakes on K8sDatapathConfig test.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
docs: Use '--address' options in kubectl port-forward examples

kubectl port-forward by default binds to localhost addresses
(127.0.0.1 and ::1, for IPv4 and IPv6, respectively). This is not
sufficient if kubectl runs in a VM from which the port is forwarded to
the host. To allow for such VM to host port forwarding, use explicit
'--address' options with addresses '0.0.0.0' and '::' to make the port
available on all IPs of the node.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
test: Forward grafana port from k8s1 vagrant VM to the host

Forward the grafana port (3000) from k8s1 VM to the host so that
Grafana may be accessed from a host browser. For this to work, kubectl
port-forward must be running within k8s1, e.g.:

  kubectl -n cilium-monitoring port-forward service/grafana --address 0.0.0.0 --address :: 3000:3000

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

fristonio commented on test/k8sT/StressPolicy.go:
```diff
@@ -0,0 +1,120 @@
+	})
+
+	AfterAll(func() {
+		kubectl.CloseSSHClient()
```
It would be great if we can clean up the installed cilium for this context in `AfterAll` block. Just to make sure we have consistency within each `Describe`.

`deploymentManager.DeleteCilium()`

jrajahalme replied:
I thought this would have also cleaned up Cilium:
```
	AfterEach(func() {
		deploymentManager.DeleteAll()
	})
```
but apparently not, thanks!

jrajahalme replied:
Fixed :-)

fristonio reviewed (changes requested):
Looks good, just a minor nit. :shipit:

aanm commented on test/k8sT/manifests/etcd-deployment.yaml:
```diff
@@ -4,7 +4,7 @@ metadata:
   name: "stateless-etcd"
 spec:
   ports:
-  - port: 2379
```
This can still be `2379`

jrajahalme replied:
@aanm I had the test failing due to port collision with the port 2379 whenever `stateless-etcd` and `etcd-k8s1` were deployed on the same node (`k8s1`). Would appreciate elaboration on how this can still be 2379? Force deployment onto `k8s2` instead?

jrajahalme replied:
Or do you mean that the `spec` can say `port: 2379` even if the actual listening port is different due to the aforementioned conflict with `etcd-k8s1`?

aanm replied:
> Or do you mean that the `spec` can say `port: 2379` even if the actual listening port is different due to the aforementioned conflict with `etcd-k8s1`?

Yes, this

jrajahalme replied:
@aanm Isn't it still better to have the spec have the port we actually use?

jrajahalme replied:
So I don't get the benefit of leaving the port as 2379...

aanm replied:
@jrajahalme there's no major benefit, I was just pointing out that this port number does not reflect the port number used on lines 44 and 45 in this file.

aanm commented on test/k8sT/manifests/policy-stress-test.yaml:
```diff
@@ -0,0 +1,1066 @@
+# Based on an automatically generated connectivity-check.
+# Edited manually.
```
Small nit, it's not possible to generate this file? I'm just thinking this file might get outdated.

jrajahalme replied:
For now, not possible, edits have been made locally. Will have to resolve later, if necessary.

jrajahalme replied:
In general, the L7 policies in the connectivity-check were not elaborate enough to be useful for the stress test. Also had to add a label to satisfy the test manifest abstraction, but then used that same label in half of the generated identities and the CNPs as well. Will have to see if these changes are acceptable as part of examples/kubernetes/connectivity-check. Current test manifest helper requires yamls be placed within `test/k8sT/manifests/`, though.

fristonio reviewed (approved):
LGTM :rocket:

christarazi commented on test/k8sT/StressPolicy.go:
```diff
@@ -0,0 +1,121 @@
+
+			// Insert 20000 identities, half matching a fromEndpoints in policy, half matching any wildcarded fromEndpoints
+			ciliumPod, err := kubectl.GetCiliumPodOnNodeWithLabel(helpers.CiliumNamespace, helpers.K8s1)
+			ExpectWithOffset(1, err).Should(BeNil(), "Unable to determine cilium pod on node %s", helpers.K8s1)
```
This should be `Expect` without an offset.

jrajahalme replied:
Fixed :-)

christarazi commented on test/k8sT/StressPolicy.go:
```diff
@@ -0,0 +1,121 @@
+	newlineRegexp := regexp.MustCompile(`\n[ \t\n]*`)
+	trimNewlines := func(script string) string {
+		return newlineRegexp.ReplaceAllLiteralString(script, " ")
+	}
```
Is it necessary to have this? Could we format our strings to avoid using something like this instead?

jrajahalme replied:
Not really as the script is really simple. Fixed.

christarazi reviewed (changes requested):
LGTM overall, just a few minor comments.

joestringer reviewed (approved):
I only looked at docs changes, LGTM.

tklauser commented on test/k8sT/StressPolicy.go:
```diff
@@ -0,0 +1,121 @@
+	})
+
+	AfterFailed(func() {
+		kubectl.CiliumReport(helpers.CiliumNamespace,
+			"cilium endpoint list")
```
`kubectl.CiliumReport` was changed in #13107 to no longer take the namespace argument:

```suggestion
		kubectl.CiliumReport("cilium endpoint list")
```

jrajahalme replied:
Fixed, thanks!

tklauser commented on test/k8sT/StressPolicy.go:
```diff
@@ -0,0 +1,121 @@
+			deploymentManager.WaitUntilReady()
+
+			// Insert 20000 identities, half matching a fromEndpoints in policy, half matching any wildcarded fromEndpoints
+			ciliumPod, err := kubectl.GetCiliumPodOnNodeWithLabel(helpers.CiliumNamespace, helpers.K8s1)
```
Same here for `kubectl.GetCiliumPodOnNodeWithLabel`:

```suggestion
			ciliumPod, err := kubectl.GetCiliumPodOnNodeWithLabel(helpers.K8s1)
```

jrajahalme replied:
Fixed, thanks!

tklauser reviewed (commented):
Two nits regarding changed test helper funcs which show up after rebasing to latest `master`.

jrajahalme committed:
test: Add a nightly policy stress test

Add a nightly policy stress test using the connectivity-check as the
base deployment, but using CNPs that exercise the policy computation
while thousands of security identitites are added to the cluster via
`cilium kvstore` commands.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

tklauser commented on Documentation/troubleshooting.rst:
```diff
@@ -275,11 +275,11 @@ service by port-forwarding it:
 This will forward the Hubble Relay service port (``80``) to your local machine
-on port ``4245``. The next step consists of downloading the latest binary
-release of Hubble CLI from the
+on port ``4245`` on all of it's IP addresses. The next step consists of
```
Nit:

```suggestion
on port ``4245`` on all of its IP addresses. The next step consists of
```

tklauser reviewed (approved):
One small nit in the troubleshooting docs, not worth respinning the PR to fix that.
